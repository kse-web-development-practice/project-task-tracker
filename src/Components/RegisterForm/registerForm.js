import React, { useState, useContext } from 'react'
import PropTypes from 'prop-types'
import styles from './registerForm.module.css'
import { FormInput } from '../FormInput/formInput'
import { Button } from '../Button/button'
import { UserContext } from '../../user-context'
import { useNavigate } from 'react-router-dom'

export const RegisterForm = ({ registerFunc }) => {
  const [username, setUsername] = useState('')
  const [password, setPassword] = useState('')
  const [repeatPassword, setRepeatPassword] = useState('')
  const [error, setError] = useState('')

  const userContext = useContext(UserContext)

  const navigate = useNavigate()

  const handleSubmit = (event) => {
    event.preventDefault()
    if (!username || !password) {
      setError('Please fill in all fields.')
      return
    }
    const user = registerFunc(username, password)
    if (!user) {
      setError('User already exist.')
      return
    }
    userContext.setUser(user.username, user.token)
    navigate('/')
    setError('')
  }

  return (
    <form className={styles.registerForm} onSubmit={handleSubmit}>
      {error && <div className="error">{error}</div>}
      <FormInput
        labelText="Username:"
        type=""
        value={username}
        onChange={(e) => setUsername(e.target.value)}
      />
      <FormInput
        labelText="Password:"
        type="password"
        value={password}
        onChange={(e) => setPassword(e.target.value)}
      />
      <FormInput
        labelText="Repeat password:"
        type="password"
        value={repeatPassword}
        onChange={(e) => setRepeatPassword(e.target.value)}
      />
      <Button isMain={true} type="submit">
        Register
      </Button>
    </form>
  )
}

RegisterForm.propTypes = {
  registerFunc: PropTypes.func
}
