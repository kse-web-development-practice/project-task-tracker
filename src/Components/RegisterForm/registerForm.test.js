import React from 'react'
import { render, screen, fireEvent, act, waitFor } from '@testing-library/react'
import { RegisterForm } from './registerForm'
import { useNavigate } from 'react-router-dom'
import { UserContext } from '../../user-context'

jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn()
}))

describe('RegisterForm Component', () => {
  const renderRegisterForm = (registerFunc) =>
    render(
      <UserContext.Provider value={{ setUser: jest.fn() }}>
        <RegisterForm registerFunc={registerFunc} />
      </UserContext.Provider>
    )

  test('default snapshot', () => {
    const { asFragment } = renderRegisterForm()
    expect(asFragment()).toMatchSnapshot()
  })

  test('empty fields error snapshot', () => {
    const { asFragment } = renderRegisterForm()

    fireEvent.click(screen.getByText('Register'))

    expect(asFragment()).toMatchSnapshot()
  })

  test('passwords not match error snapshot', async () => {
    const { asFragment } = renderRegisterForm()
    act(() => {
      fireEvent.change(screen.getByLabelText(/Username:/i), { target: { value: 'testUser' } })
      fireEvent.change(screen.getByTestId('password-input'), { target: { value: 'password123' } })
      fireEvent.change(screen.getByTestId('repeat-password-input'), {
        target: { value: 'password321' }
      })

      fireEvent.click(screen.getByText('Register'))
    })

    expect(asFragment()).toMatchSnapshot()
  })

  test('valid submission', async () => {
    const registerFunc = jest.fn().mockResolvedValue({ username: 'testUser', token: 'abc123' })
    const mockNavigate = jest.fn()
    useNavigate.mockReturnValue(mockNavigate)

    renderRegisterForm(registerFunc)
    act(() => {
      fireEvent.change(screen.getByLabelText(/Username:/i), { target: { value: 'testUser' } })
      fireEvent.change(screen.getByTestId('password-input'), { target: { value: 'password123' } })
      fireEvent.change(screen.getByTestId('repeat-password-input'), {
        target: { value: 'password123' }
      })

      fireEvent.click(screen.getByText('Register'))
    })

    await waitFor(() => {
      expect(registerFunc).toHaveBeenCalledWith('testUser', 'password123')

      expect(mockNavigate).toHaveBeenCalledWith('/')
    })
  })
})
