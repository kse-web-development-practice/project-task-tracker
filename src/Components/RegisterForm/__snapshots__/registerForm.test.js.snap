// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm Component default snapshot 1`] = `
<DocumentFragment>
  <form
    class="registerForm"
  >
    <label
      class="formLabel"
    >
      Username:
      <input
        class="formInput"
        type=""
        value=""
      />
    </label>
    <label
      class="formLabel"
    >
      Password:
      <input
        class="formInput"
        data-testid="password-input"
        type="password"
        value=""
      />
    </label>
    <label
      class="formLabel"
    >
      Repeat password:
      <input
        class="formInput"
        data-testid="repeat-password-input"
        type="password"
        value=""
      />
    </label>
    <button
      class="button main"
    >
      Register
    </button>
  </form>
</DocumentFragment>
`;

exports[`RegisterForm Component empty fields error snapshot 1`] = `
<DocumentFragment>
  <form
    class="registerForm"
  >
    <div
      class="error"
    >
      Please fill in all fields.
    </div>
    <label
      class="formLabel"
    >
      Username:
      <input
        class="formInput"
        type=""
        value=""
      />
    </label>
    <label
      class="formLabel"
    >
      Password:
      <input
        class="formInput"
        data-testid="password-input"
        type="password"
        value=""
      />
    </label>
    <label
      class="formLabel"
    >
      Repeat password:
      <input
        class="formInput"
        data-testid="repeat-password-input"
        type="password"
        value=""
      />
    </label>
    <button
      class="button main"
    >
      Register
    </button>
  </form>
</DocumentFragment>
`;

exports[`RegisterForm Component passwords not match error snapshot 1`] = `
<DocumentFragment>
  <form
    class="registerForm"
  >
    <div
      class="error"
    >
      Incorrect password repeat.
    </div>
    <label
      class="formLabel"
    >
      Username:
      <input
        class="formInput"
        type=""
        value="testUser"
      />
    </label>
    <label
      class="formLabel"
    >
      Password:
      <input
        class="formInput"
        data-testid="password-input"
        type="password"
        value="password123"
      />
    </label>
    <label
      class="formLabel"
    >
      Repeat password:
      <input
        class="formInput"
        data-testid="repeat-password-input"
        type="password"
        value="password321"
      />
    </label>
    <button
      class="button main"
    >
      Register
    </button>
  </form>
</DocumentFragment>
`;
